### PR TITLE
fix for PLFM-2079

### PR DESCRIPTION
**Pull request by john-hill**

### CODE REVIEW

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/web/controller/JSONEntityHttpMessageConverter.java [hunk 1]
```diff
@@ -80,15 +84,33 @@ public boolean canWrite(Class<?> clazz, MediaType mediaType) {
...
+					jsonObject.put(CONCRETE_TYPE, type);
+					jsonString = jsonObject.toString();
+					// try again
+					return EntityFactory.createEntityFromJSONString(jsonString, clazz);
+				}else{
+					// Something else went wrong
+					throw new HttpMessageNotReadableException(e.getMessage());
+				}
+			} catch (JSONException e1) {
+				throw new HttpMessageNotReadableException(e.getMessage());
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/web/controller/JSONEntityHttpMessageConverter.java]: do you mean e.getMessage() or e1.getMessage()?

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/web/controller/JSONEntityHttpMessageConverter.java [hunk 2]
```diff
@@ -80,15 +84,33 @@ public boolean canWrite(Class<?> clazz, MediaType mediaType) {
...
+					// get the entity type so we can replace it with concrete type
+					String type = jsonObject.getString(ENTITY_TYPE);
+					jsonObject.remove(ENTITY_TYPE);
+					jsonObject.put(CONCRETE_TYPE, type);
+					jsonString = jsonObject.toString();
+					// try again
+					return EntityFactory.createEntityFromJSONString(jsonString, clazz);
+				}else{
+					// Something else went wrong
+					throw new HttpMessageNotReadableException(e.getMessage());
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/web/controller/JSONEntityHttpMessageConverter.java]: can you use exception chaining here, i.e. to capture 'e'?  (same for two exception constructors below)
throw new HttpMessageNotReadableException(e.getMessage(), e);

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/web/controller/JSONEntityHttpMessageConverterTest.java [hunk 3]
```diff
@@ -162,4 +168,37 @@ public void testCreateEntityFromeAdapterBadJSON() throws JSONObjectAdapterExcept
...
+		JSONObject jsonObject = new JSONObject();
+		JSONObjectAdapter adapter = new JSONObjectAdapterImpl(jsonObject);
+		project.writeToJSONObject(adapter);
+		// Swap the concreteType field with entityType
+		String type = jsonObject.getString("concreteType");
+		jsonObject.remove("concreteType");
+		// replace it with entity type
+		jsonObject.put("entityType", type);
+		String json = adapter.toJSONString();
+		System.out.println(json);
```
Comment (brucehoff) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/web/controller/JSONEntityHttpMessageConverterTest.java]: can you remove the print statement?

Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/web/controller/JSONEntityHttpMessageConverterTest.java]: no

**Pull request merged**